tsfc/driver.py

```diff
@@ -243,3 +250,30 @@ def facetarea():
 
     kernel_name = "%s_%s_integral_%s" % (prefix, integral_type, integral_data.subdomain_id)
     return builder.construct_kernel(kernel_name, body)
+
+
+def lower_integral_type(fiat_cell, integral_type):
+    dim = fiat_cell.get_dimension()
+    if integral_type == 'cell':
+        integration_dim = dim
+    elif integral_type in ['exterior_facet', 'interior_facet']:
+        integration_dim = dim - 1
+    else:
+        basedim, extrdim = dim
+        assert extrdim == 1
+
+        if integral_type in ['exterior_facet_vert', 'interior_facet_vert']:
+            integration_dim = (basedim - 1, 1)
+        elif integral_type in ['exterior_facet_bottom', 'exterior_facet_top', 'interior_facet_horiz']:
+            integration_dim = (basedim, 0)
+        else:
+            raise NotImplementedError("integral type %s not supported" % integral_type)
+
+    if integral_type == 'exterior_facet_bottom':
+        entity_ids = [0]
+    elif integral_type == 'exterior_facet_top':
+        entity_ids = [1]
+    else:
+        entity_ids = range(len(fiat_cell.get_topology()[integration_dim]))
+
+    return integration_dim, entity_ids
```

Comment:
Not sure what you mean by `integration_entity`. Let's say if we are on a triangle, then `integration_dim` is 2 for cell integrals, and 1 for facet integrals, rather than a triangle cell for cell integrals and an interval cell for facet integrals.

I guess I'm missing a docstring here anyway.